onnxruntime/core/framework/utils.h

```diff
@@ -87,69 +87,399 @@ void DumpNodeInputs(const OpKernelContext& context, const Node& node);
+      retval = function<uint64_t>(__VA_ARGS__);                            \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_FLOAT16:                     \
+      retval = function<MLFloat16>(__VA_ARGS__);                           \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16:                    \
+      retval = function<BFloat16>(__VA_ARGS__);                            \
+      break;                                                               \
+    default:                                                               \
+      ORT_ENFORCE(false, "Unknown tensor type of ", tensor_type);          \
+  }
+
+////////////////////////////////////////////////////////////////////////////////
+/// Use the following primitives if you have a few types to switch on so you
+//  can write a short sequence of if/else statements.
+
+// This is a frequently used check so we make a separate utility function.
+inline bool IsDataTypeString(MLDataType dt_type) {
+  auto prim_type = dt_type->AsPrimitiveDataType();
+  return (prim_type != nullptr && prim_type->GetDataType() == ONNX_NAMESPACE::TensorProto_DataType_STRING);
+}
+
+// Test if MLDataType is a concrete type of PrimitiveDataTypeBase
+// and it is T
+template <class T>
+inline bool IsPrimitiveDataType(MLDataType dt_type) {
+  auto prim_type = dt_type->AsPrimitiveDataType();
+  return (prim_type != nullptr && prim_type->GetDataType() == ToTensorProtoElementType<T>());
+}
+
+// Use after AsPrimitiveDataType() is successful
+// Check if PrimitiveDataTypeBase is of type T
+template <class T>
+inline bool IsPrimitiveDataType(const PrimitiveDataTypeBase* prim_type) {
+  assert(prim_type != nullptr);
+  return prim_type->GetDataType() == ToTensorProtoElementType<T>();
+}
+
+// This implementation contains a workaround for GCC bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=47226
+// GCC until very recently does not support template parameter pack expansion within lambda context.
+namespace mltype_dispatcher_internal {
+// T - type handled by this helper
+struct CallableDispatchableHelper {
+  int32_t dt_type_;  // Type currently dispatched
+  size_t called_;
+
+  explicit CallableDispatchableHelper(int32_t dt_type) : dt_type_(dt_type), called_(0) {}
+
+  // Must return integer to be in a expandable context
+  template <class T, class Fn, class... Args>
+  int Invoke(Fn&& fn, Args&&... args) {
+    if (utils::ToTensorProtoElementType<T>() == dt_type_) {
+      std::forward<Fn>(fn)(std::forward<Args>(args)...);
+      ++called_;
+    }
+    return 0;
+  }
+};
+
+// Default policy is to throw with no return type.
+template<class Ret>
+struct UnsupportedTypeDefaultPolicy {
+  Ret operator()(int32_t dt_type) const {
+    ORT_THROW("Unsupported data type: ", dt_type);
+    return Ret();
```

Comment:
is the return statement after the throw needed?

Reply:
No idea. I wish the compilers would follow the same standard.

---
In reply to: [343300340](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r343300340) [](ancestors = 343300340)

Reply:
Done.

---
In reply to: [343908352](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r343908352) [](ancestors = 343908352,343300340)